File: src/main/java/com/dambra/paul/stringcalculator/RecursivelyProcessBracketedCalculations.java
package com.dambra.paul.stringcalculator;

import java.util.ArrayList;

class RecursivelyProcessBracketedCalculations {
    static ArrayList<String> from(ArrayList<String> parts) {
        if (!parts.contains("(")) {
            return parts;
        }

        var start = parts.indexOf("(");
        var end = parts.indexOf(")");
        var insideBrackets = String.join("", parts.subList(start + 1, end));

        var result = new StringCalculator().calculate(insideBrackets);

        var x = replaceBracketsDelimitedOperationsWithResult(parts, start, end, result);
        return RecursivelyProcessBracketedCalculations.from(x);
    }

    private static ArrayList<String> replaceBracketsDelimitedOperationsWithResult(ArrayList<String> parts, int start, int end, String result) {
        var x = new ArrayList<String>();
        for (var i = 0; i < parts.size(); i++) {
            if (i < start || end < i) {
                x.add(parts.get(i));
            }
            if (i == end) {
                x.add(result);
            }
        }
        return x;
    }
}

File: src/main/java/com/dambra/paul/stringcalculator/RecursivelyProcessBracketedCalculations.kt
package com.dambra.paul.stringcalculator

internal object RecursivelyProcessBracketedCalculations {
    fun from(parts: MutableList<String>): MutableList<String> {
        if (!parts.contains("(")) {
            return parts
        }

        val start = parts.indexOf("(")
        val end = parts.indexOf(")")

        val result = calculateInsideBrackets(parts, start, end)

        val filteredParts = replaceBracketsWithResult(parts, start, end, result)

        return RecursivelyProcessBracketedCalculations.from(filteredParts)
    }

    private fun replaceBracketsWithResult(parts: MutableList<String>, start: Int, end: Int, result: String): MutableList<String> {
        val filteredParts = parts.filterIndexed { i, _ -> i < start || i > end }.toMutableList()
        filteredParts.add(start, result)
        return filteredParts
    }

    private fun calculateInsideBrackets(parts: MutableList<String>, start: Int, end: Int) =
            StringCalculator().calculate(getContentsOfBrackets(parts, start, end))

    private fun getContentsOfBrackets(parts: MutableList<String>, start: Int, end: Int) =
            parts.subList(start + 1, end).joinToString("")

}
